qiskit/quantum_info/synthesis/one_qubit_decompose.py

```diff
@@ -14,233 +14,393 @@
-        # NOTE: The 4th variable is phase to be used later
-        theta, phi, lam, _ = self._angles(unitary_mat)
-        if self._basis == 'U3':
-            return self._circuit_u3(theta, phi, lam)
-        if self._basis == 'U1X':
-            return self._circuit_u1x(theta,
-                                     phi,
-                                     lam,
-                                     simplify=simplify,
-                                     atol=atol)
-        if self._basis == 'ZYZ':
-            return self._circuit_zyz(theta,
-                                     phi,
-                                     lam,
-                                     simplify=simplify,
-                                     atol=atol)
-        if self._basis == 'ZXZ':
-            return self._circuit_zxz(theta,
-                                     phi,
-                                     lam,
-                                     simplify=simplify,
-                                     atol=atol)
-        if self._basis == 'XYX':
-            return self._circuit_xyx(theta,
-                                     phi,
-                                     lam,
-                                     simplify=simplify,
-                                     atol=atol)
-        raise QiskitError("OneQubitEulerDecomposer: invalid basis")
+    @property
+    def basis(self):
+        """The decomposition basis."""
+        return self._basis
+
+    @basis.setter
+    def basis(self, basis):
+        """Set the decomposition basis."""
+        basis_methods = {
+            'U3': (self._params_u3, self._circuit_u3),
+            'U1X': (self._params_u1x, self._circuit_u1x),
+            'RR': (self._params_zyz, self._circuit_rr),
+            'ZYZ': (self._params_zyz, self._circuit_zyz),
+            'ZXZ': (self._params_zxz, self._circuit_zxz),
+            'XYX': (self._params_xyx, self._circuit_xyx)
+        }
+        if basis not in basis_methods:
+            raise QiskitError("OneQubitEulerDecomposer: unsupported basis {}".format(basis))
+        self._basis = basis
+        self._params, self._circuit = basis_methods[self._basis]
+
+    def angles(self, unitary):
+        """Return the Euler angles for input array.
+
+        Args:
+            unitary (np.ndarray): 2x2 unitary matrix.
+
+        Returns:
+            tuple: (theta, phi, lambda).
+        """
+        theta, phi, lam, _ = self._params(unitary)
+        return theta, phi, lam
+
+    def angles_and_phase(self, unitary):
+        """Return the Euler angles and phase for input array.
+
+        Args:
+            unitary (np.ndarray): 2x2 unitary matrix.
+
+        Returns:
+            tuple: (theta, phi, lambda, phase).
+        """
+        return self._params(unitary)
+
+    def circuit(self, theta, phi, lam, simplify=True,
+                simplify_tolerance=DEFAULT_SIMPLIFY_TOLERANCE):
+        """Return the basis circuit for the input parameters.
```

Comment:
Why is there both a public and a private `circuit()`/`_circuit()` method which do the same thing? This public one doesn't seem to be used anywhere so can you delete it?

Reply:
The private one is set to the current basis circuit value `_circuit_u3` or `_circuit_zyz`

Reply:
but where is the public one used?

Reply:
I guess it isn't currently used, but idea was it could be used to return the decomposition circuit for fixed parameters without doing the synthesis from a matrix. ie `decomposer.circuit(theta, phi, lam)`. However all the current classes that use the decomposer only use the angles and then build the circuit/definition themselves. I can remove it if you want.

Reply:
Yes I'd say it should be removed. Getting theta, phi, lam themselves can be done using the U3 decomposition